File: app/src/main/java/jp/hazuki/yuzubrowser/action/ActionFile.java
package jp.hazuki.yuzubrowser.action;

import android.content.Context;

import com.fasterxml.jackson.core.JsonFactory;
import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.Serializable;

import jp.hazuki.yuzubrowser.utils.ErrorReport;
import jp.hazuki.yuzubrowser.utils.Logger;

public abstract class ActionFile implements Serializable {
    private static final long serialVersionUID = 9159377694255234638L;

    private static final String TAG = "ActionFile";

    public abstract File getFile(Context context);

    public abstract void reset();

    public boolean load(Context context) {
        reset();

        File file = getFile(context);
        if (!file.exists() || file.isDirectory()) return true;

        try (InputStream is = new BufferedInputStream(new FileInputStream(file))) {

            JsonFactory factory = new JsonFactory();
            JsonParser parser = factory.createParser(is);

            if (!load(parser)) {
                Logger.e(TAG, "loadMain error (return false)");
                parser.close();
                return false;
            }

            parser.close();
            return true;
        } catch (IOException e) {
            ErrorReport.printAndWriteLog(e);
        }
        return false;
    }

    public abstract boolean load(JsonParser parser) throws IOException;

    public boolean write(Context context) {
        File file = getFile(context);

        try (OutputStream os = new BufferedOutputStream(new FileOutputStream(file))) {

            JsonFactory factory = new JsonFactory();
            JsonGenerator generator = factory.createGenerator(os);

            if (!write(generator)) {
                Logger.e(TAG, "writeMain error (return false)");
                generator.close();
                return false;
            }

            generator.close();
            return true;
        } catch (IOException e) {
            ErrorReport.printAndWriteLog(e);
        }
        return false;
    }

    public abstract boolean write(JsonGenerator generator) throws IOException;
}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/action/ActionFile.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.action

import android.content.Context
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import jp.hazuki.yuzubrowser.core.utility.log.ErrorReport
import jp.hazuki.yuzubrowser.core.utility.log.Logger
import okio.Okio
import okio.buffer
import okio.sink
import okio.source
import java.io.File
import java.io.IOException
import java.io.Serializable

abstract class ActionFile : Serializable {

    abstract fun getFile(context: Context): File

    abstract fun reset()

    fun load(context: Context): Boolean {
        reset()

        val file = getFile(context)
        if (!file.exists() || file.isDirectory) return true

        try {
            JsonReader.of(file.source().buffer()).use {
                if (!load(it)) {
                    Logger.e(TAG, "loadMain error (return false)")
                    return false
                }
                return true
            }
        } catch (e: IOException) {
            ErrorReport.printAndWriteLog(e)
        }

        return false
    }

    @Throws(IOException::class)
    abstract fun load(reader: JsonReader): Boolean

    fun write(context: Context): Boolean {
        val file = getFile(context)

        try {
            JsonWriter.of(file.sink().buffer()).use {
                if (!write(it)) {
                    Logger.e(TAG, "writeMain error (return false)")
                    return false
                }
                it.flush()
                return true
            }
        } catch (e: IOException) {
            ErrorReport.printAndWriteLog(e)
        }
        return false
    }

    @Throws(IOException::class)
    abstract fun write(writer: JsonWriter): Boolean

    companion object {
        private const val serialVersionUID = 9159377694255234638L

        private const val TAG = "ActionFile"
    }
}
